content/tutorials/reactjs-realtime-commenting.textile

```diff
@@ -0,0 +1,344 @@
+      <section className="section">
+        {
+          this.props.comments.map((comment, index) => {
+            return <Comment key={index} comment={comment} />
+          })
+        }
+      </section>
+    );
+  }
+}
+
+export default Comments;
+```
+
+This component accepts a @comments@ props and renders the Comment component for each of the comments available. It passes the actual comment to the Comment component as props.
+
+h2. Step 12 - Create App Component
+
+The App component will server as the parent component. Open @components/App.js@ and replace it with the code below:
+
+```[javascript]
+import React, { Component } from 'react';
+import CommentBox from './CommentBox';
+import Comments from './Comments';
+import Ably from '../ably';
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+        comments: []
+    }
+  }
+
+  render() {
+    return (
+      <section className="section">
+        <div className="container">
+          <div className="columns">
+            <div className="column is-half is-offset-one-quarter">
+              <CommentBox />
+              <Comments comments={this.state.comments} />
+            </div>
+          </div>
+        </div>
+      </section>
+    );
+  }
+}
+
+export default App;
+```
+
+This contains the components we created earlier. We define a @comments@ state which is an array of comments. It defaults to an empty array. This will be updated as we add comment through the comment form. Also, we pass the @comments@ state as props to the Comments component. This is how the Comments component get the comments it renders.
+
+h2. Step 13 - Subscribing to Comments
+
+We have seen how we can publish newly added comments. Now, we need to subscribe to those comments being published. We’ll do this within React lifecycle @componentDidMount()@ hook. Still within @components/App.js@, add the code below just before the @render()@ function:
+
+```[javascript]
+componentDidMount() {
+  // Subscribe to comments
+  const channel = Ably.channels.get('comments');
+
+  channel.subscribe('add_comment', (comment) => {
+    this.setState((prevState) => {
+      return {
+        comments: prevState.comments.concat({
+          name: comment.data.name,
+          comment: comment.data.comment
+        })
+      };
+    });
+  });
+}
+```
+
+The @componentDidMount()@ hook runs after the component output has been rendered to the DOM. It is a good place to subscribe to comments. We subscribe to the @comments@ channel and listen for the @add_comment@ event. We then update the @comments@ state with the comment published from the channel we are subscribed to.
+
+<a href="/images/tutorials/reactjs-realtime-commenting-final-output.png" target="_blank">
+  <img src="/images/tutorials/reactjs-realtime-commenting-final-output.png" style="width: 100%" alt="Final output screenshot">
+</a>
+
+That’s it. We now have a realtime commenting system. To test it out, start the app:
```

Comment:
Should we not use history to prepopulate the comments?

For example, if you add the following to `componentDidMount`:

```js
channel.attach();
    channel.once('attached', () => {
      channel.history((err, page) => {
        // page.items contains list of up to 100 previous items
      });
    });
```

See https://www.ably.io/tutorials/history#lang-javascript, you could create a similar step to Step 2, and then even link to the history tutorial (with `./history` as the path) if the user wants to understand how to use history.

Reply:
Wasn't aware of history before.

Reply:
Yeah, using history can be one way. I think it might not really suit the implementation of this tutorial. I'm adding the newly created comment to the `state`, that way, we can do whatsoever we like with the data in `state` like persisting to a database.

Reply:
> I think it might not really suit the implementation of this tutorial.

But the point of the tutorial is to expose Ably functionality, and without a server in this tutorial, this is an obvious way to solve two problems.